From 3679d010e1875bfc66f4b86592516d74bede21f5 Mon Sep 17 00:00:00 2001
From: Khem Raj <raj.khem@gmail.com>
Date: Fri, 21 Nov 2025 17:39:07 -0800
Subject: [PATCH] filan: Fix build with glibc C23 strchr

glibc master with C23-enabled headers makes strchr() const-correct, so
when called with a const char * it now returns const char *. `printtime()`
stored the result of asctime() in a `const char *s` and then did:

    if (strchr(s, '\n')) *strchr(s, '\n') = '\0';

With the new prototype this fails to compile with:

    error: read-only variable is not assignable

as the result of strchr() is treated as pointing to read-only memory.

asctime() already returns a mutable `char *`, so fix the function by
using a `char *s` and assigning the result of strchr() to a temporary
pointer before writing the terminating NUL. This restores the previous
behaviour and allows socat to build against latest glibc master while
remaining correct on older releases as well.

Upstream-Status: Submitted [sent to socat@dest-unreach.org]
Signed-off-by: Khem Raj <raj.khem@gmail.com>
---
 filan.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/filan.c b/filan.c
index fb9229c..3c1e376 100644
--- a/filan.c
+++ b/filan.c
@@ -1055,14 +1055,16 @@ const char *getfiletypestring(int st_mode) {
 }

 static int printtime(FILE *outfile, time_t time) {
-   const char *s;
+   char *s;

    if (filan_rawoutput) {
       fprintf(outfile, "\t"F_time, time);
    } else {
       fputc('\t', outfile);
       s = asctime(localtime(&time));
-      if (strchr(s, '\n'))  *strchr(s, '\n') = '\0';
+      char *nl = strchr(s, '\n');
+      if (nl)
+        *nl = '\0';
       fputs(s, outfile);
    }
    return 0;
